File: rl-designer/src/services/collection.ts
import { invoke } from '@tauri-apps/api/core';
import type { DecalTextures } from '@/types';
import type { ElementType } from '@/constants/elements';

interface FetchResult {
  success: boolean;
  decals: DecalTextures[];
  error?: string;
}

export const getDecalFolders = async (element: ElementType): Promise<FetchResult> => {
  try {
    const result = await invoke<FetchResult>(`get_element_decal_texture_folder`, { element });
    return result;
  } catch (error) {
    console.error(`Failed to fetch ${element} decal folders:`, error);
    return {
      success: false,
      decals: [],
      error: `Failed to invoke command: ${error}`
    };
  }
}

interface RemoveResult {
  success: boolean;
  error?: string;
}

interface RemoveDecalVariantParams {
  element: ElementType;
  decalName: string;
  variantName: string;
}

export const removeDecalVariant = async ({element, decalName, variantName}: RemoveDecalVariantParams): Promise<RemoveResult> => {
  try {
    const result = await invoke<RemoveResult>(`remove_element_decal_variant`, { element, decalName, variantName });
    return result;
  } catch (error) {
    console.error(`Failed to remove ${element} decal variant:`, error);
    return {
      success: false,
      error: `Failed to invoke command: ${error}`
    };
  }
};

File: rl-designer/src-tauri/src/commands/collection.rs
use crate::types::decal::DecalInfo;
use crate::types::elements::ElementType;
use crate::utils::collection::{fetch_decal_folders, remove_decal_variant_logic};

#[derive(Debug, serde::Serialize)]
pub struct FetchResult {
    success: bool,
    decals: Vec<DecalInfo>,
    error: Option<String>,
}

#[tauri::command]
pub fn get_element_decal_texture_folder(element: ElementType) -> FetchResult {
    match fetch_decal_folders(element) {
        Ok(decals) => FetchResult {
            success: true,
            decals,
            error: None,
        },
        Err(error) => FetchResult {
            success: false,
            decals: vec![],
            error: Some(error),
        },
    }
}

#[derive(Debug, serde::Serialize)]
pub struct RemoveResult {
    success: bool,
    error: Option<String>,
}

#[tauri::command]
pub fn remove_element_decal_variant(
    element: ElementType,
    decal_name: String,
    variant_name: String,
) -> RemoveResult {
    match remove_decal_variant_logic(element, &decal_name, &variant_name) {
        Ok(_) => RemoveResult {
            success: true,
            error: None,
        },
        Err(error) => RemoveResult {
            success: false,
            error: Some(error),
        },
    }
}
